pkg/R/sparkR.R

```diff
@@ -116,16 +116,33 @@ sparkR.init <- function(
   if (yarn_conf_dir != "") {
     cp <- paste(cp, yarn_conf_dir, sep = ":")
   }
-  launchBackend(classPath = cp,
-                mainClass = "edu.berkeley.cs.amplab.sparkr.SparkRBackend",
-                args = as.character(sparkRBackendPort),
-                javaOpts = paste("-Xmx", sparkMem, sep = ""))
+
+  sparkRExistingPort <- Sys.getenv("EXISTING_SPARKR_BACKEND_PORT", "")
+  if (sparkRExistingPort != "") {
+    sparkRBackendPort <- sparkRExistingPort
+  } else {
+    if (Sys.getenv("SPARKR_USE_SPARK_SUBMIT", "") == "") {
+      launchBackend(classPath = cp,
+                    mainClass = "edu.berkeley.cs.amplab.sparkr.SparkRBackend",
+                    args = as.character(sparkRBackendPort),
+                    javaOpts = paste("-Xmx", sparkMem, sep = ""))
+    } else {
+      # TODO: We should deprecate sparkJars and ask users to add it to the
+      # command line (using --jars) which is picked up by SparkSubmit
+      launchBackendSparkSubmit(
+          mainClass = "edu.berkeley.cs.amplab.sparkr.SparkRBackend",
+          args = as.character(sparkRBackendPort),
+          appJar = .sparkREnv$assemblyJarPath,
+          sparkHome = sparkHome,
+          sparkSubmitOpts = Sys.getenv("SPARKR_SUBMIT_ARGS", ""))
+    }
+  }
 
   .sparkREnv$sparkRBackendPort <- sparkRBackendPort
   cat("Waiting for JVM to come up...\n")
   tries <- 0
-  while(tries < sparkRRetryCount) {
-    if(!connExists(.sparkREnv)) {
+  while (tries < sparkRRetryCount) {
+    if (!connExists(.sparkREnv)) {
       Sys.sleep(2 ^ tries)
```

Comment:
With the merge from master, we are going to sleep for at least 1 second when we are simply reconnect to backend on the existing port (sparkRExistingPort path on line 122).

It is likely fine as we need to try/retry the connection even in that case. But the sleep might not be absolutely necessary, since it is not launching a new process. Thought?